// appium-flutter-finder/build/serializer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var base64url_1 = require("./base64url");
var deserializer_1 = require("./deserializer");
// @todo consider using protobuf
function serialize(obj) {
    return base64url_1.encode(JSON.stringify(obj));
}
exports.ancestor = function (args) {
    var of = args.of, matching = args.matching, _a = args.matchRoot, matchRoot = _a === void 0 ? false : _a;
    var a = {
        finderType: "Ancestor",
        matchRoot: matchRoot,
    };
    Object.entries(deserializer_1.deserialize(of)).forEach(function (_a) {
        var key = _a[0], value = _a[1];
        return (a["of_" + key] = value);
    });
    Object.entries(deserializer_1.deserialize(matching)).forEach(function (_a) {
        var key = _a[0], value = _a[1];
        return (a["matching_" + key] = value);
    });
    return serialize(a);
};
exports.bySemanticsLabel = function (label) {
    return serialize({
        finderType: "BySemanticsLabel",
        isRegExp: label instanceof RegExp ? true : false,
        label: label instanceof RegExp ? label.toString().slice(1, -1) : label,
    });
};
exports.byTooltip = function (text) {
    return serialize({
        finderType: "ByTooltipMessage",
        text: text,
    });
};
exports.byType = function (type) {
    return serialize({
        finderType: "ByType",
        type: type,
    });
};
exports.byValueKey = function (key) {
    return serialize({
        finderType: "ByValueKey",
        keyValueString: key,
        keyValueType: typeof key === "string" ? "String" : "int",
    });
};
exports.descendant = function (args) {
    var of = args.of, matching = args.matching, _a = args.matchRoot, matchRoot = _a === void 0 ? false : _a;
    var a = {
        finderType: "Descendant",
        matchRoot: matchRoot,
    };
    Object.entries(deserializer_1.deserialize(of)).forEach(function (_a) {
        var key = _a[0], value = _a[1];
        return (a["of_" + key] = value);
    });
    Object.entries(deserializer_1.deserialize(matching)).forEach(function (_a) {
        var key = _a[0], value = _a[1];
        return (a["matching_" + key] = value);
    });
    return serialize(a);
};
exports.pageBack = function () {
    return serialize({
        finderType: "PageBack",
    });
};
exports.byText = function (text) {
    return serialize({
        finderType: "ByText",
        text: text,
    });
};
